modules/renter/repair.go

```diff
@@ -26,323 +25,374 @@ var renewThreshold = func() types.BlockHeight {
+type chunkGaps struct {
+	contracts []types.FileContractID
+	pieces    []uint64
+	numGaps   int
 }
 
-type hostErrs []*hostErr
-
-func (hs hostErrs) Error() string {
-	var errs []error
-	for _, h := range hs {
-		errs = append(errs, h)
+func (r *Renter) addFileToRepairMatrix(file *file, availableWorkers map[types.FileContractID]struct{}, repairMatrix map[string]*chunkGaps, gapCounts map[int]int) {
+	// Flatten availableWorkers into a list of contracts.
+	contracts := make([]types.FileContractID, 0)
+	for contract := range availableWorkers {
+		contracts = append(contracts, contract)
 	}
-	return build.JoinErrors(errs, "\n").Error()
-}
 
-// repair attempts to repair a file chunk by uploading its pieces to more
-// hosts.
-func (f *file) repair(chunkIndex uint64, missingPieces []uint64, r io.ReaderAt, hosts []contractor.Editor) error {
-	// read chunk data and encode
-	chunk := make([]byte, f.chunkSize())
-	_, err := r.ReadAt(chunk, int64(chunkIndex*f.chunkSize()))
-	if err != nil && err != io.EOF && err != io.ErrUnexpectedEOF {
-		return err
+	// If the file is not being tracked, don't add it to the repair matrix.
+	_, exists := r.tracking[file.name]
+	if exists {
+		return
 	}
-	pieces, err := f.erasureCode.Encode(chunk)
-	if err != nil {
-		return err
+
+	// For each chunk, create a map from the chunk to the pieces that it
+	// has, and to the contracts that have that chunk.
+	chunkCount := file.numChunks()
+	contractMap := make(map[uint64][]types.FileContractID)
+	pieceMap := make(map[uint64][]uint64)
+	for i := uint64(0); i < chunkCount; i++ {
+		contractMap[i] = make([]types.FileContractID, 0)
+		pieceMap[i] = make([]uint64, 0)
 	}
-	// encrypt pieces
-	for i := range pieces {
-		key := deriveKey(f.masterKey, chunkIndex, uint64(i))
-		pieces[i], err = key.EncryptBytes(pieces[i])
-		if err != nil {
-			return err
+
+	// Iterate through each contract and figure out what's available.
+	for _, contract := range file.contracts {
+		for _, piece := range contract.Pieces {
+			contractMap[piece.Chunk] = append(contractMap[piece.Chunk], contract.ID)
+			pieceMap[piece.Chunk] = append(pieceMap[piece.Chunk], piece.Piece)
 		}
 	}
 
-	// upload one piece per host
-	numPieces := len(missingPieces)
-	if len(hosts) < numPieces {
-		numPieces = len(hosts)
-	}
-	errChan := make(chan *hostErr)
-	for i := 0; i < numPieces; i++ {
-		go func(pieceIndex uint64, host contractor.Editor) {
-			// upload data to host
-			root, err := host.Upload(pieces[pieceIndex])
-			if err != nil {
-				errChan <- &hostErr{host.Address(), err}
-				return
+	// Iterate through each chunk and, if there are gaps, add the inverse
+	// to the repair matrix.
+	for i := uint64(0); i < chunkCount; i++ {
+		if len(pieceMap[i]) < file.erasureCode.NumPieces() {
+			// Find the gaps in the pieces and contracts.
+			potentialPieceGaps := make(map[uint64]struct{})
+			for j := 0; j < file.erasureCode.NumPieces(); i++ {
+				potentialPieceGaps[i] = struct{}{}
+			}
```

Comment:
you got your `i` and `j` mixed up.

`make([]bool, file.erasureCode.NumPieces())` might be a better data structure for this purpose.